445 offline workflow supervisors

jacobvjk opened the pull request:
closes #445 

This PR:

- adds an alternative second step to the offline PACTA workflow
- this step can be used to prepare the portfolio data whenever the user needs to run PACTA offline on their own machines and cannot grant us access to any of their data
- It requires different input data from our side (and the user's side!), as described [here](https://docs.google.com/document/d/1p-k4M1-ku8y6UIsFrZLhAdUzNoZDcMiLn36RIAxZkoY/edit#heading=h.7fb5m5qd2r64)
- It leads to a new workflow, described in detail [here](https://docs.google.com/document/d/1rBMLBOtQJGwoQVGjy6qRH71hiv9XFeXMI86eE50yunQ/edit#)
- introduces a boolean that indicates whether the script is to be run remotely or not
- runs checks and definitions of column types based on that boolean as they differ between the workflows.

Note:

- I am absolutely open to discussing the implementation of this workflow, as some of the choices I made are rather for reasons of getting this out soon than for thoroughly thought through reasons regarding future restructuring of the package.

jacobvjk commented:
re R CMD checks on mac: `Error: Failed to get R 4.0.5`

maurolepore commented:
> re R CMD checks on mac: Error: Failed to get R 4.0.5

https://github.com/2DegreesInvesting/r2dii.plot.static/pull/76#pullrequestreview-633809317

maurolepore reviewed (approved):
LGTM but I'm not a good person to judge this code -- 'm not familiar with it and and have no way to easily see the impact of the change (e.g. reproducible example or test). 

I'm inclined to approve without much investigation because:

* based on the PR's first comment -- this changes do not impact the user-facing "online" tool so it feels quite safe. Also, having the

* I think we should have only one code base and the one it's now public is the so called "online" tool. So I suspect that changes to the "offline" tool won't affect the code in the long term.

The only general comment I can make is about the commended code. Why not remove it?

jacobvjk commented:
@maurolepore this is the part I am open to talking about. I don't remove it, because this works specifically for the new remote offline version. In principle, I think the files could be adjusted in the normal offline version as well to all fit the same logic, but that involves a few loops with AR and data_preparation. The main difference being that I have replaced bbg IDs in some of the data sets with AR's company IDs, because we cannot share bbg IDs. The type difference is that bbg IDs are mixed character and numbers and therefore the id column was cast as character, whereas the AR id is numeric. so in order to avoid issues with a join, I removed this. The following check regarding col types could probably be kept/adjusted as well, but since there is no documentation about the col names that refer to the order of these col types, it's hard to know haw it would have to be adjusted. Maybe @cjyetman can also chime in on this?

cjyetman commented:
@jacobvjk Can we have a short chat about this today? Honestly, it pains me to see yet another separate workflow for PACTA rather than one universal one, but I don't want to make an unnecessary blocker for you.

jacobvjk commented:
@cjyetman absolutely, I'm by no means tied to doing it this way. It's really more a suggestion that I'm hoping to turn into something that won't cause further headaches

cjyetman commented:
FYI... macOS CI is working again... I reran the jobs so it's passing now

jacobvjk commented:
so I added another boolean in the initialisation of the project in script 1. Based on that the user either should run the standard script 2 or 2b. For now they are separate files, but if we find consensus that this should rather all happen in script 2 and be decided by the boolean as well, then I am happy to merge that into one script. In script 3, column types and checks depend on the boolean as well. fyi @cjyetman

maurolepore commented:
I'm happy for you to merge whenever you are ready, based on your conversation with CJ and personal judgement.
